subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -691,43 +692,42 @@ private boolean doesConsumerDependOnDestroyer(Node consumer, Node destroyer) {
     }
 
     private static boolean hasOverlap(Iterable<String> paths1, Iterable<String> paths2) {
-        for (String path1 : paths1) {
-            for (String path2 : paths2) {
-                String overLappedPath = getOverLappedPath(path1, path2);
-                if (overLappedPath != null) {
-                    return true;
-                }
-            }
+        //build a helper TreeSet with the contents of paths2
+        final TreeSet<String> set = new TreeSet<>();
```

Comment:
I suppose a `FileHierarchySet` would be better suited to check whether or not the paths overlap. I wonder if it is faster to create a `FileHierarchySet` from the smaller or larger set of paths.

Reply:
I now tried to implement an algorithm based on (Default)FileHierarchySet, but I could not get my unit tests to run: FileHierarchySet is a File-based API, and I only have strings. I tried to adapt to the API by wrapping the String into a File, but DefaultFileHierarchySet calls getAbsolutePath() in plus() and getPath() in contains(), whichs leads to an incompatible behavior for relative paths. Furthermore, I doubt that the performance will be adequate (at least for my tests), because DefaultFileHierarchySet still does a linear scan over all children of a directory.

Which way shall I go:
1. change the implementation of FileHierarchySet so that it works as needed and give it another try,
2. change the unit tests and accept that the semantics of the algorithm will change for certain paths, or
3. stay with the TreeSet based implementation?